# ai03bTasks
# Machine Learning: Decision Trees
## Data Cleaning

**Instructions:**
- Complete each task below by running the code cells
- Fill in the blanks and answer questions in markdown cells
- Save your work when finished
- Push this file to your GitHub "Machine Learning" Repo under the appropriate folder.

---
## Setup: Import Libraries and Load Data

Run this cell first to set up your environment.

In [17]:
import pandas as pd

# Load the Titanic dataset if it's not already
df = pd.read_csv("Titanic Dataset.csv")

print("✓ Data loaded successfully!")
print(f"Original shape: {df.shape}")
print(f"\nOriginal columns: {df.columns.tolist()}")

✓ Data loaded successfully!
Original shape: (1309, 14)

Original columns: ['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket', 'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest']


---
## Task 1: Understand the Original Data

Before cleaning, let's see what we're working with.

### 1a. How many rows and columns are in the original dataset?

In [18]:
# TODO: Print the shape of the DataFrame
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

Rows: 1309
Columns: 14


### 1b. Display the first few rows

In [19]:
# TODO: Use .head() to display the first 5 rows
print("First 5 rows:")
print(df.head())

First 5 rows:
   pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     S

### 1c. Check for missing values in the original data

In [20]:
# TODO: Use .isnull().sum() to count missing values per column
print("Missing values in original data:")
print(df.isnull().sum())

Missing values in original data:
pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64


**Q: Which columns have the most missing values?**

A: body

---
## Task 2: Select Useful Features

We'll keep only the columns that help predict survival.

### 2a. Keep only these 8 columns: pclass, survived, sex, age, sibsp, parch, fare, embarked

In [21]:
# TODO: Select only the useful columns
# Hint: df = df[['column1', 'column2', ...]]
df = df[['pclass', 'survived', 'sex', 'age', 
         'sibsp', 'parch', 'fare', 'embarked']]

print("✓ Columns selected!")
print(f"New shape: {df.shape}")
print(f"\nColumns: {df.columns.tolist()}")

✓ Columns selected!
New shape: (1309, 8)

Columns: ['pclass', 'survived', 'sex', 'age', 'sibsp', 'parch', 'fare', 'embarked']


### 2b. Explain why we dropped certain columns

**Q: Why did we drop the 'name' column?**.  (Refer to the lesson if you don't recall - or ask a neighbor!)

A: It isn't important

**Q: Why did we drop the 'cabin' column?**

A: because almost everybody has a different cabin, it doesn't help

**Q: Why did we drop the 'boat' column?**

A: because it doesn't effect if they survived

---
## Task 3: Check for Missing Values

Now let's see which of our selected columns have missing values.

In [22]:
# TODO: Check for missing values in the cleaned dataset
print("Missing values after feature selection:")
print(df.isnull().sum())

Missing values after feature selection:
pclass        0
survived      0
sex           0
age         263
sibsp         0
parch         0
fare          1
embarked      2
dtype: int64


**Q: How many missing values are in the 'age' column?**

A: 263

**Q: How many missing values are in the 'fare' column?**

A: 1

**Q: How many missing values are in the 'embarked' column?**

A: 2

---
## Task 4: Handle Missing Age Values

Age has many missing values. We'll fill them with the median age.

### 4a. Calculate the median age

In [23]:
# TODO: Calculate the median age
median_age = df['age'].median()
print(f"Median age: {median_age}")

Median age: 28.0


### 4b. Fill missing ages with the median

In [24]:
# TODO: Fill missing ages with median_age
# Hint: Use .fillna(value, inplace=True)
df['age'].fillna(median_age, inplace=True)

print("✓ Missing ages filled with median!")
print(f"Missing ages now: {df['age'].isnull().sum()}")

✓ Missing ages filled with median!
Missing ages now: 0


### 4c. Compare median vs mean for age

In [25]:
# Calculate both median and mean
print(f"Median age: {df['age'].median():.2f}")
print(f"Mean age: {df['age'].mean():.2f}")

Median age: 28.00
Mean age: 29.50


**Q: Which is larger, the median or mean? Why might this be?**

A: mean, it includes the outliers

---
## Task 5: Handle Missing Fare Values

Fare has only 1 missing value. We'll fill it with the median fare.

In [26]:
# TODO: Calculate median fare
median_fare = df['fare'].median()
print(f"Median fare: ${median_fare:.2f}")

# TODO: Fill missing fare with median
df['fare'].fillna(median_fare, inplace=True)

print("✓ Missing fare filled!")
print(f"Missing fares now: {df['fare'].isnull().sum()}")

Median fare: $14.45
✓ Missing fare filled!
Missing fares now: 0


---
## Task 6: Handle Missing Embarked Values

Embarked has only 2 missing values. Since this is so few, we'll drop those rows.

### 6a. How many rows before dropping?

In [27]:
rows_before = len(df)
print(f"Rows before dropping: {rows_before}")

Rows before dropping: 1309


### 6b. Drop rows with missing embarked values

In [28]:
# TODO: Drop rows where 'embarked' is missing
# Hint: Use .dropna(subset=['column_name'], inplace=True) - also in the lesson material
df.dropna(subset=['embarked'], inplace=True)

rows_after = len(df)
rows_dropped = rows_before - rows_after

print("✓ Rows with missing embarked dropped!")
print(f"Rows after dropping: {rows_after}")
print(f"Rows dropped: {rows_dropped}")
print(f"Missing embarked now: {df['embarked'].isnull().sum()}")

✓ Rows with missing embarked dropped!
Rows after dropping: 1307
Rows dropped: 2
Missing embarked now: 0


**Q: What percentage of data did we lose by dropping these rows?**

A: 0.001%

---
## Task 7: Verify All Missing Values Are Gone

Let's do a final check to make sure our data is completely clean.

In [29]:
# TODO: Check for any remaining missing values
print("Final missing value check:")
print(df.isnull().sum())
print(f"\nTotal missing values: {df.isnull().sum().sum()}")

Final missing value check:
pclass      0
survived    0
sex         0
age         0
sibsp       0
parch       0
fare        0
embarked    0
dtype: int64

Total missing values: 0


**Q: Are there any missing values remaining? (Should be 0!)**

A: 0

---
## Task 8: Summary Statistics

Now that our data is clean, let's look at summary statistics.

In [30]:
# Display summary statistics
print("Summary statistics after cleaning:")
df.describe()

Summary statistics after cleaning:


pclass     survived          age        sibsp        parch  \
count  1307.000000  1307.000000  1307.000000  1307.000000  1307.000000   
mean      2.296863     0.381025    29.471821     0.499617     0.385616   
std       0.836942     0.485825    12.881592     1.042273     0.866092   
min       1.000000     0.000000     0.170000     0.000000     0.000000   
25%       2.000000     0.000000    22.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    35.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare  
count  1307.000000  
mean     33.209595  
std      51.748768  
min       0.000000  
25%       7.895800  
50%      14.454200  
75%      31.275000  
max     512.329200

**Q: What is the average age after filling missing values?**

A: 29.471821

**Q: What is the average fare?**

A: 33.209595

---
## Task 9: Save the Cleaned Data
Save your cleaned data to a new CSV file so you can use it in the next lesson.

In [31]:
# Save cleaned data
# Writes the DataFrame to a new CSV file without adding extra row numbers (index)
df.to_csv("Titanic_Cleaned.csv", index=False)     

# Prints a confirmation message so the user knows the save was successful
print("✓ Cleaned data saved to 'Titanic_Cleaned.csv'")   

✓ Cleaned data saved to 'Titanic_Cleaned.csv'


---
## Reflection Questions

Answer these questions based on your work:

**1. Why is it important to check for missing values before building a model?**

Answer: it could cause outliers

**2. When should you fill missing values vs. drop rows?**

Answer: you should only fill values if there is a significant amount missing

**3. Why did we use median instead of mean to fill missing ages?**

Answer: because you dont want it to be effected by outliers

**4. What could happen if we trained a model on data with missing values?**

Answer: it could cause mistakes

**5. Name one real-world scenario where missing data might occur.**

Answer: if you have a corrupted drive and you lose some of your data

---
## Lesson Complete!

You've successfully cleaned the Titanic dataset!

**Summary of what you did:**
- Selected 8 useful features from 15 columns
- Filled missing ages with median
- Filled missing fares with median
- Dropped 2 rows with missing embarked values
- Verified all missing values are gone

Save this notebook and push to GitHub.

**Next lesson**: Convert categorical data to numbers